packages/@aws-cdk/core/lib/arn.ts

```diff
@@ -75,12 +75,33 @@ export class Arn {
    * The required ARN pieces that are omitted will be taken from the stack that
    * the 'scope' is attached to. If all ARN pieces are supplied, the supplied scope
    * can be 'undefined'.
+   * @param components the components of the ARN to be formatted
+   * @param stack the Stack, used to pull Arn components if any of 'partition',
+   *              'region', or 'account' are undefined
    */
-  public static format(components: ArnComponents, stack: Stack): string {
-    const partition = components.partition !== undefined ? components.partition : stack.partition;
-    const region = components.region !== undefined ? components.region : stack.region;
-    const account = components.account !== undefined ? components.account : stack.account;
-    const sep = components.sep !== undefined ? components.sep : '/';
+  public static format(components: ArnComponents, stack?: Stack): string {
+    const requiredComponents: Array<keyof ArnComponents> = ['partition', 'region', 'account'];
+    let definedComponents: Required<Omit<ArnComponents, 'resourceName' | 'sep'>>;
+    if (!stack) {
+      for (const rc of requiredComponents) {
+        if (components[rc] === undefined) {
+          throw new Error(`Stack must be defined if "${rc}" ARN component is undefined`);
+        }
+      }
+      definedComponents = components as Required<ArnComponents>;
+    } else {
+      definedComponents = {
+        ...components,
+        partition: components.partition ?? stack.partition,
+        region: components.region ?? stack.region,
+        account: components.account ?? stack.account,
+      };
+    }
```

Comment:
This is unnecessarily convoluted and hard to reason about. 

Any reason we can't use this solution:

```ts
const componentRequiredError = c => new Error(`"stack" must be defined since "${c}" is not defined`);
if (components.partition == null) { throw componentRequiredError('partition'); }
if (components.region == null)    { throw componentRequiredError('region');    }
if (components.account == null)   { throw componentRequiredError('account');   }
```

(using `== null` to capture both `null` and `undefined`).